### DISCUSSION TITLE
How to control the volume of `AudioSourceSPIFFS` and `AudioSourceURL`?

### DISCUSSION THREAD

**Discussion by AlphaGaming7780** [pschatzmann/arduino-audio-tools#1714] [Q&A] [answered]
I'm trying to make the volume working with `AudioSourceSPIFFS` and `AudioSourceURL`, but nothing work.

For SPIFFS I have followed the example (https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-spiffs-i2s/player-spiffs-i2s.ino) but the audioPlayer.setVolume() does nothing, I tried with multiple value (like 0.5f, 0.1f, 0.01f) nothing change.

I have the same result with the `AudioSourceURL`.

The output is the I2SStream to an external DAC.

Does somebody have any idea on how I can make this work ?

**Comment by pschatzmann** [comment]
I never had any issues with this and there were not changes recently which could cause any such issues.
Does it make a difference if you set the volume before or after calling begin ?
If you set the log level to info: do you see that the volume gets updated ?

**Reply by AlphaGaming7780** [reply]
Thanks for the quick response, I tried and no change, the volume remains the same whether before or after.

**Comment by pschatzmann** [comment, marked as answer]
Just made a test and I could not reproduce your issue:  
- player.setVolume(0.0); was producing silence
- player.setVolume(0.1); was quite low volume
- player.setVolume(1.0); was quite loud

both when called before or after player.begin().

if you look at your library.properties what version number do you see ?
Do you have maybe an i2s module with a dynamic volume control which would automatically amplify a lower input ?

Here are the relevant log entries for the volume of 0.1
```
...
[I] StreamCopy.h : 70 - buffer_size=1024
[I] AudioPlayer.h : 338 - setVolume(0.100000)
[I] VolumeStream.h : 199 - setVolume: 0.100000 at 0
[I] VolumeStream.h : 199 - setVolume: 0.100000 at 1
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioTypes.h : 128 - MP3DecoderHelix sample_rate: 48000 / channels: 2 / bits_per_sample: 16
[I] AudioPlayer.h : 230 - sample_rate: 48000
[I] AudioPlayer.h : 231 - bits_per_sample: 16
[I] AudioPlayer.h : 232 - channels: 2
[I] AudioTypes.h : 128 - in: sample_rate: 48000 / channels: 2 / bits_per_sample: 16
[I] AudioTypes.h : 128 - out: sample_rate: 48000 / channels: 2 / bits_per_sample: 16
[I] I2SCodecStream.h : 93 - virtual void audio_tools::I2SCodecStream::setAudioInfo(audio_tools::AudioInfo)
[I] AudioTypes.h : 128 - in: sample_rate: 48000 / channels: 2 / bits_per_sample: 16
[I] AudioTypes.h : 128 - out: sample_rate: 48000 / channels: 2 / bits_per_sample: 16
[I] I2SStream.h : 93 - virtual void audio_tools::I2SStream::setAudioInfo(audio_tools::AudioInfo)
[I] AudioTypes.h : 128 - in: sample_rate: 48000 / channels: 2 / bits_per_sample: 16
[I] AudioTypes.h : 128 - out: sample_rate: 48000 / channels: 2 / bits_per_sample: 16
[I] I2SESP32V1.h : 40 - i2s_set_sample_rates: 48000
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
...
```

**Reply by AlphaGaming7780** [reply]
version=0.9.8

> Do you have maybe an i2s module with a dynamic volume control which would automatically amplify a lower input ?

Nop, I used Bluetooth lib before and the volume can be 0 with my phone.

How can I show all these logs? Maybe I'm going to find something useful in it.

**Comment by pschatzmann** [comment]
Just set the [log level to Info](https://github.com/pschatzmann/arduino-audio-tools/wiki/Introduction#logging)

**Reply by AlphaGaming7780** [reply]
Nothing strange here, IDK, I will try more stuff and maybe find out.
```
[I] AudioPlayer.h : 338 - setVolume(0.100000)
[I] VolumeStream.h : 199 - setVolume: 0.100000 at 0
[I] VolumeStream.h : 199 - setVolume: 0.100000 at 1
[I] AudioTypes.h : 128 - in: sample_rate: 41000 / channels: 2 / bits_per_sample: 16
[I] AudioTypes.h : 128 - out: sample_rate: 41000 / channels: 2 / bits_per_sample: 16
[I] AudioTypes.h : 128 -  sample_rate: 41000 / channels: 2 / bits_per_sample: 16
[I] I2SConfigESP32.h : 80 - rx/tx mode: TX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 41000
[I] I2SConfigESP32.h : 84 - bits per sample: 16
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_LSB_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 97 - buffer_count:8
[I] I2SConfigESP32.h : 98 - buffer_size:64
[I] I2SConfigESP32.h : 103 - pin_bck: 15
[I] I2SConfigESP32.h : 105 - pin_ws: 14
[I] I2SConfigESP32.h : 107 - pin_data: 22
[I] AudioPlayer.h : 404 - virtual void audio_tools::AudioPlayer::setMetadataCallback(void (*)(audio_tools::MetaDataType, const char*, int), audio_tools::ID3TypeSelection)
[I] AudioPlayer.h : 338 - setVolume(0.010000)
[I] VolumeStream.h : 199 - setVolume: 0.010000 at 0
[I] VolumeStream.h : 199 - setVolume: 0.010000 at 1
[I] AudioTypes.h : 128 - in: sample_rate: 41000 / channels: 2 / bits_per_sample: 16
[I] AudioTypes.h : 128 - out: sample_rate: 41000 / channels: 2 / bits_per_sample: 16
[I] MetaDataID3.h : 569 - virtual void audio_tools::MetaDataID3::begin()
[I] AudioSourceSPIFFS.h : 49 - selectStream: 0
[I] AudioSourceSPIFFS.h : 53 - Using file /Portal Radio.mp3
[I] StreamCopy.h : 70 - buffer_size=1024
Portal Radio started!
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
```

**Reply by pschatzmann** [reply]
I was just checking in the code: Can you double check in the AudioPlayer.h on line 135 if isResultPCM() return true ? (e.g. by adding some trace or log statements). For non pcm data the volume control can not work and that would explain your symptoms

I committed a correction that might help, if this was the issue and it returned false...

**Reply by AlphaGaming7780** [reply]
I tried by adding this
![image](https://github.com/user-attachments/assets/8b844be1-1733-4f05-bdf9-06d308ff49cb)

And got that: 
![image](https://github.com/user-attachments/assets/7ec2f38c-7a07-40ec-a9a5-694bcd48e1b3)

From what you said, this is correct. I just think my weird code create that “bug”, just need to find why.

**Reply by AlphaGaming7780** [reply]
Okay, I have done a test without the rest of my project and the volume work fine.

Can the issue come from here ? Because I'm using `setAudioSource` ?
![image](https://github.com/user-attachments/assets/dc2356ba-f2b7-45c9-8461-2c3e272cc175)

EDIT: Yeah, I made these changes in my test environment:
- Comments the previous declaration of player (which was working fine)
![image](https://github.com/user-attachments/assets/76012b26-195b-4437-980b-94d6e85c36bb)

- And set the source, the output, and the decoder using these lines:
![image](https://github.com/user-attachments/assets/9f3d00d5-cf34-4dcd-8f5f-7b4c9eb9ddd6)

**Reply by pschatzmann** [reply]
Make sure that you define the Output after setDecoder()

**Reply by AlphaGaming7780** [reply]
I try, the volume is still max.
![image](https://github.com/user-attachments/assets/403b44fa-bd8c-4529-b43d-66cd67b29b68)

**Reply by pschatzmann** [reply]
What is the data type of i2SOutput.GetAudioStream() ?
Maybe the volume_out has never been started: does it make a difference if you add volume_out.begin() in the AudioPlayer begin method ?

**Reply by AlphaGaming7780** [reply]
I create this class just to store all my config.
```c++
    I2SStream _stream

    audio_tools::AudioStream &GetAudioStream() 
    {
        return _stream;
    }

    void begin() {

        auto config = _stream.defaultConfig(TX_MODE);
        config.i2s_format = I2S_LSB_FORMAT;
        config.pin_ws = 14;
        config.pin_bck = 15;
        config.pin_data = 22;
        config.sample_rate = 41000;
        config.bits_per_sample = I2S_BITS_PER_SAMPLE_16BIT;
        config.channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT;
        config.buffer_count = 8;
        config.buffer_size = 64;
        config.use_apll = false;
        config.auto_clear = true; // avoiding noise in case of data unavailability
        _stream.begin(config);
    }
```

**Comment by pschatzmann** [comment]
Ok, here is my understanding: you also need to add the following 
```c++
    decoder.addNotifyAudioChange(player);
```
So that the player is notifed about the format by the decoder. This is with the current logic also setting up the VolumeStream volume_out.

I committed a correction that is calling begin() on the volume_out which should make the issue go away as well. But without notification you run the risk that the sample rate, channels and bits per sample are not matching with the actual played back audio

**Reply by AlphaGaming7780** [reply]
That setup works, thank you for the time and the help.
![image](https://github.com/user-attachments/assets/0d9283e4-adb7-4b0f-93e4-2c090780918a)

**Reply by AlphaGaming7780** [reply]
I still have an issue with the URL stream the audio stop randomly every fixed time and after few min he randomly times out 
(When I disable the logs, it doesn't change anything.) (I used that example : https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-communication/http-client/player-url-i2s/player-url-i2s.ino)
![image](https://github.com/user-attachments/assets/8b504b35-d36b-48e4-a68f-b6085b5d42de)

**Comment by pschatzmann** [comment]
What ESP32 core version are you using ?

On my first thought there might be an issue with the Wifi or internet connection: Did you try to bring the ESP32 closer to the router or try another radio station that is in your country ?

You can try to adjust the timeout parameters in the AudioSource ...

**Reply by AlphaGaming7780** [reply]
Hey, sorry for the long wait, I made few tests and found the web radio I used is 196kb/s mp3, after changing to 128kbs/s it fixed the issue.
Do you know a wait to fix the issue of using 192kb/s ?

**Comment by pschatzmann** [comment]
Try to find out if the issue is caused by the decoder or by the WiFi!
If it is the decoder you can play around with the memory settings or try another decoder...
